File: python/src/laguerre.py
"""
Generalized (associated) Laguerre polynomial L_n^alpha(x) via the standard
three-term forward recurrence. Matches MATLAB's laguerreL(n, alpha, x).

    L_0^alpha(x) = 1
    L_1^alpha(x) = 1 + alpha - x
    L_{k+1}^alpha(x) = ((2k + 1 + alpha - x) * L_k - (k + alpha) * L_{k-1}) / (k + 1)
"""

import numba


@numba.njit(cache=True)
def laguerre_L(n, alpha, x):
    """Generalized Laguerre polynomial L_n^alpha(x) via three-term recurrence.

    Parameters
    ----------
    n : int
        Polynomial degree (>= 0).
    alpha : int
        Generalized order.
    x : float
        Evaluation point.

    Returns
    -------
    float
        Value of L_n^alpha(x).
    """
    if n == 0:
        return 1.0
    L_prev = 1.0
    L_curr = 1.0 + alpha - x
    if n == 1:
        return L_curr
    for k in range(1, n):
        L_next = ((2 * k + 1 + alpha - x) * L_curr - (k + alpha) * L_prev) / (k + 1)
        L_prev = L_curr
        L_curr = L_next
    return L_curr

File: rust/src/laguerre.rs
/// Generalized Laguerre polynomial L_n^alpha(x) via three-term recurrence.
pub fn laguerre_l(n: i32, alpha: i32, x: f64) -> f64 {
    if n <= 0 {
        return 1.0;
    }

    let mut l_prev = 1.0;
    let mut l_curr = 1.0 + alpha as f64 - x;
    if n == 1 {
        return l_curr;
    }

    for k in 1..n {
        let kf = k as f64;
        let af = alpha as f64;
        let l_next = ((2.0 * kf + 1.0 + af - x) * l_curr - (kf + af) * l_prev) / (kf + 1.0);
        l_prev = l_curr;
        l_curr = l_next;
    }

    l_curr
}
